modules/apps/27-interchain-accounts/host/client/cli/tx_test.go

```diff
@@ -0,0 +1,127 @@
+
+const bankSendMessage = `{
+	"@type":"/cosmos.bank.v1beta1.MsgSend",
+	"from_address":"cosmos15ccshhmp0gsx29qpqq6g4zmltnnvgmyu9ueuadh9y2nc5zj0szls5gtddz",
+	"to_address":"cosmos10h9stc5v6ntgeygf5xf945njqq5h32r53uquvw",
+	"amount": [
+		{
+			"denom": "stake",
+			"amount": "1000"
+		}
+	]
+}`
+
+func TestGeneratePacketData(t *testing.T) {
+	tests := []struct {
+		name                string
+		memo                string
+		expectedPass        bool
+		message             string
+		registerInterfaceFn func(registry codectypes.InterfaceRegistry)
+		assertionFn         func(msg sdk.Msg)
+	}{
+		{
+			name:                "packet data generation succeeds (MsgDelegate)",
+			memo:                "non-empty-memo",
+			expectedPass:        true,
+			message:             msgDelegateMessage,
+			registerInterfaceFn: stakingtypes.RegisterInterfaces,
+			assertionFn: func(msg sdk.Msg) {
+				msgDelegate, ok := msg.(*stakingtypes.MsgDelegate)
+				require.True(t, ok)
+				require.Equal(t, "cosmos15ccshhmp0gsx29qpqq6g4zmltnnvgmyu9ueuadh9y2nc5zj0szls5gtddz", msgDelegate.DelegatorAddress)
+				require.Equal(t, "cosmosvaloper1qnk2n4nlkpw9xfqntladh74w6ujtulwnmxnh3k", msgDelegate.ValidatorAddress)
+				require.Equal(t, "stake", msgDelegate.Amount.Denom)
+				require.Equal(t, uint64(1000), msgDelegate.Amount.Amount.Uint64())
+			},
+		},
+		{
+			name:                "packet data generation succeeds (MsgSend)",
+			memo:                "non-empty-memo",
+			expectedPass:        true,
+			message:             bankSendMessage,
+			registerInterfaceFn: banktypes.RegisterInterfaces,
+			assertionFn: func(msg sdk.Msg) {
+				bankSendMsg, ok := msg.(*banktypes.MsgSend)
+				require.True(t, ok)
+				require.Equal(t, "cosmos15ccshhmp0gsx29qpqq6g4zmltnnvgmyu9ueuadh9y2nc5zj0szls5gtddz", bankSendMsg.FromAddress)
+				require.Equal(t, "cosmos10h9stc5v6ntgeygf5xf945njqq5h32r53uquvw", bankSendMsg.ToAddress)
+				require.Equal(t, "stake", bankSendMsg.Amount.GetDenomByIndex(0))
+				require.Equal(t, uint64(1000), bankSendMsg.Amount[0].Amount.Uint64())
+			},
+		},
+		{
+			name:                "empty memo is valid",
+			memo:                "",
+			expectedPass:        true,
+			message:             msgDelegateMessage,
+			registerInterfaceFn: stakingtypes.RegisterInterfaces,
+			assertionFn:         func(msg sdk.Msg) {},
+		},
+		{
+			name:                "invalid message string",
+			expectedPass:        false,
+			message:             "<invalid-message-body>",
+			registerInterfaceFn: func(codectypes.InterfaceRegistry) {},
+			assertionFn:         func(sdk.Msg) {},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		ir := codectypes.NewInterfaceRegistry()
+		tt.registerInterfaceFn(ir)
+
+		cdc := codec.NewProtoCodec(ir)
+		t.Run(tt.name, func(t *testing.T) {
+			bz, err := generatePacketData(cdc, []byte(tt.message), tt.memo)
+			if !tt.expectedPass {
+				require.Error(t, err)
+				require.Nil(t, bz)
+				return
+			}
+			require.NoError(t, err)
+			require.NotNil(t, bz)
```

Comment:
typically we do `tc.expectedPass` first, is there a benefit to changing this behaviour?

Reply:
The only reason I opted for this way is to keep the code as flat as possible, i.e. return on failure rather than have if/else. I'm happy to do the passing case first to stay consistent with the rest of the test code though.

Reply:
Makes sense! I have no preference here, I'd be happy to head in a different direction going forward if we like the flat structure better